### PR TITLE
SIP-4: DWN Routes

### PR DESCRIPTION
**Pull request by nitro-neal**
New DWN route for publishing a manifest, with dwn config for dwn endpoint

Input:
{{baseUrl}}/v1/dwn/manifests

```
{
  "manifestId": "WA-DL-CLASS-A"
}

```
Output
```
credentialManifest: {...}
published: true
```

image

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -157,6 +163,10 @@ func LoadConfig(path string) (*SSIServiceConfig, error) {
 				BaseServiceConfig:  &BaseServiceConfig{Name: "keystore"},
 				ServiceKeyPassword: "default-password",
 			},
+			DWNConfig: DWNServiceConfig{
+				BaseServiceConfig: &BaseServiceConfig{Name: "did"},
```
Comment (decentralgabe) [inline config/config.go:167]:
```suggestion
				BaseServiceConfig: &BaseServiceConfig{Name: "dwn"},
```

File: pkg/server/framework/request.go [hunk 2]
```diff
@@ -95,3 +96,20 @@ func Decode(r *http.Request, val interface{}) error {
 
 	return nil
 }
+
+// Post does a post request with data to provided endpoint
+func Post(endpoint string, data interface{}) (*http.Response, error) {
+	// convert response payload to json
+	jsonData, err := json.Marshal(data)
+	if err != nil {
```
Comment (decentralgabe) [inline pkg/server/framework/request.go:104]: consider a meaningful response to the client if they haven't provided a json request body

File: pkg/server/framework/request.go [hunk 3]
```diff
@@ -95,3 +96,20 @@ func Decode(r *http.Request, val interface{}) error {
...
+		return nil, err
+	}
+
+	resp, err := http.Post(endpoint, "application/json", bytes.NewBuffer(jsonData))
+
+	if err != nil {
```
Comment (decentralgabe) [inline pkg/server/framework/request.go:110]: here too - what kind of error would be returned to the client and logged to us?

File: pkg/server/router/dwn.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	return &DWNRouter{
+		service: dwnService,
+	}, nil
+}
+
+type DWNPostRequest struct {
+	Manifest manifest.CredentialManifest `json:"manifest" validate:"required"`
+}
+
+type PublishManifestRequest struct {
```
Comment (decentralgabe) [inline pkg/server/router/dwn.go:39]: I'm thinking this kind of request should go in a different package

maybe there should be an `dwn` package that has `requests.go` with this kind of stuff?

File: pkg/server/router/dwn.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	dwnReq := DWNPostRequest{Manifest: publishManifestResponse.Manifest}
+	dwnResp, err := framework.Post(dwnr.service.Config().DWNEndpoint, dwnReq)
+
+	if err != nil || dwnResp == nil {
+		errMsg := "problem with publishing manifest to downstream dwn"
+		logrus.WithError(err).Error(errMsg)
+		return framework.NewRequestError(errors.Wrap(err, errMsg), http.StatusBadRequest)
+	}
+
+	status := dwnResp.StatusCode >= 200 && dwnResp.StatusCode < 300
```
Comment (decentralgabe) [inline pkg/server/router/dwn.go:99]: status is a binary value?

File: pkg/server/server.go [hunk 6]
```diff
@@ -184,3 +188,15 @@ func (s *SSIServer) ManifestAPI(service svcframework.Service) (err error) {
 	s.Handle(http.MethodDelete, path.Join(responsesHandlerPath, "/:id"), manifestRouter.DeleteResponse)
 	return
 }
+
+func (s *SSIServer) DWNApi(service svcframework.Service) (err error) {
```
Comment (decentralgabe) [inline pkg/server/server.go:192]:
```suggestion
func (s *SSIServer) DWNAPI(service svcframework.Service) (err error) {
```

**Review by decentralgabe — CHANGES_REQUESTED** [review]

File: pkg/server/framework/request.go (hunk 2 above)
Comment (nitro-neal) [reply pkg/server/framework/request.go:104]: Yup I was thinking about doing this, but wanted the real response object from the dwn before I expanded upon this all the way. 

I will put in a response object on our side and guestimate what the dwn response would be

File: pkg/server/router/dwn.go (hunk 5 above)
Comment (nitro-neal) [reply pkg/server/router/dwn.go:99]: I'll fix this

File: pkg/server/router/dwn.go [hunk 7]
```diff
@@ -79,25 +75,22 @@ func (dwnr DWNRouter) PublishManifest(ctx context.Context, w http.ResponseWriter
 	}
 
 	req := request.ToServiceRequest()
-	publishManifestResponse, err := dwnr.service.PublishManifest(req)
+	publishManifestResponse, err := dwnr.service.GetManifest(req)
 
-	if err != nil {
-		errMsg := "could not publish manifest to dwn"
+	if err != nil || &publishManifestResponse.Manifest == nil {
```
Comment (decentralgabe) [inline pkg/server/router/dwn.go:80]: consider an `IsEmpty` method on `publishManifestResponse`

**Review by decentralgabe — APPROVED** [review]

**Pull request merged**